File: nodejs/selenium/pages/webform.page.ts
import { By, WebDriver, WebElement, Actions } from "selenium-webdriver";

export class WebFormPage {

  driver: WebDriver;
  input!: WebElement;
  password!: WebElement;
  textarea!: WebElement;
  selectElement!: WebElement;
  city!: WebElement;
  file!: WebElement;
  color!: WebElement;
  date!: WebElement;
  range!: WebElement;
  button!: WebElement;
  number!: SelectWrapper;

  constructor(driver: WebDriver) {
    this.driver = driver;
  }

  async init(): Promise<this> {
    this.input = await this.driver.findElement(By.name("my-text"));
    this.password = await this.driver.findElement(By.name("my-password"));
    this.textarea = await this.driver.findElement(By.name("my-textarea"));
    this.selectElement = await this.driver.findElement(By.name("my-select"));
    this.city = await this.driver.findElement(By.name("my-datalist"));
    this.file = await this.driver.findElement(By.name("my-file"));
    this.color = await this.driver.findElement(By.name("my-colors"));
    this.date = await this.driver.findElement(By.name("my-date"));
    this.range = await this.driver.findElement(By.name("my-range"));
    this.button = await this.driver.findElement(By.xpath("//button[@type='submit']"));
    this.number = new SelectWrapper(this.selectElement);

    return this;
  }

  async setInput(value: string): Promise<void> {
    await this.input.sendKeys(value);
  }

  async setPassword(value: string): Promise<void> {
    await this.password.sendKeys(value);
  }

  async setTextarea(value: string): Promise<void> {
    await this.textarea.sendKeys(value);
  }

  async setNumber(index: number): Promise<void> {
    await this.number.selectByIndex(index);
  }

  async setCity(value: string): Promise<void> {
    await this.city.sendKeys(value);
  }

  async setFile(filePath: string): Promise<void> {
    await this.file.sendKeys(filePath);
  }

  async setColor(value: string): Promise<void> {
    await this.driver.executeScript(
      "document.getElementsByName('my-colors')[0].value = arguments[0]",
      value
    );
  }

  async setDate(value: string): Promise<void> {
    await this.date.sendKeys(value);
    const body = await this.driver.findElement(By.css("body"));
    await body.click();
  }

  async setRange(value: number): Promise<void> {
    await this.driver.executeScript(
      "document.getElementsByName('my-range')[0].value = arguments[0]", value
    );
  }

  async submit(): Promise<void> {
    await this.button.click();
  }
}


class SelectWrapper {
  element: WebElement;

  constructor(element: WebElement) {
    this.element = element;
  }

  async selectByIndex(index: number): Promise<void> {
    const options = await this.element.findElements(By.tagName("option"));
    if (index < 0 || index >= options.length)
      throw new Error("Index out of range in selectByIndex");
    await options[index].click();
  }
}

File: python/selenium/pages/webform_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.remote.file_detector import LocalFileDetector


class WebFormPage:

    def __init__(self, driver):
        self.driver = driver
        self.input = driver.find_element(By.NAME, "my-text")
        self.password = driver.find_element(By.NAME, "my-password")
        self.textarea = driver.find_element(By.NAME, "my-textarea")
        self.number = Select(driver.find_element(By.NAME, "my-select"))
        self.city = driver.find_element(By.NAME, "my-datalist")
        self.driver.file_detector = LocalFileDetector()
        self.file = driver.find_element(By.NAME, "my-file")
        self.color = driver.find_element(By.NAME, "my-colors")
        self.date = driver.find_element(By.NAME, "my-date")
        self.range = driver.find_element(By.NAME, "my-range")
        self.button = driver.find_element(By.XPATH, "//button[@type='submit']")

    def set_input(self, value):
        self.input.send_keys(value)

    def set_password(self, value):
        self.password.send_keys(value)

    def set_textarea(self, value):
        self.textarea.send_keys(value)

    def set_number(self, value):
        self.number.select_by_index(value)

    def set_city(self, value):
        self.city.send_keys(value)

    def set_file(self, value):
        self.file.send_keys(value)

    def set_color(self, value):
        self.driver.execute_script(f"document.getElementsByName('my-colors')[0].value = '{value}'")

    def set_date(self, value):
        self.date.send_keys(value)
        # Click elsewhere to close the calendar
        self.driver.find_element(By.XPATH, "//body").click()

    def set_range(self, value):
        # Two solutions
        # Solution #1: Trigger mouse events to move the range handler
        dimensions = self.range.rect
        current = int(self.range.get_attribute("value"))
        step = dimensions['width'] / (int(self.range.get_attribute("max")) - int(self.range.get_attribute("min")))
        offset = (value - current) * step
        action = ActionChains(self.driver)
        action.move_to_element_with_offset(self.range, offset, 0).click().perform()

        # solution #2: Set the range value directly
        self.driver.execute_script(f"document.getElementsByName('my-range')[0].value = '{value}'")

    def submit(self):
        self.button.click()
